### PR TITLE
Fix edge cases around oversize test messages.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 1]
```diff
@@ -303,15 +303,12 @@ - (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
     if (!displayableText) {
         NSString *text = textBlock();
 
-        // Only show up to 2kb of text.
-        const NSUInteger kMaxTextDisplayLength = 2 * 1024;
-        text = [text ows_stripped];
+        // Only show up to N characters of text.
+        const NSUInteger kMaxTextDisplayLength = 1024;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: I propose we should less by default.  WDYT?

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 2]
```diff
@@ -336,70 +331,81 @@ - (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
...
-    OWSAssert([self.interaction isKindOfClass:[TSMessage class]]);
+    OWSAssert([self.interaction isKindOfClass:[TSOutgoingMessage class]] ||
+        [self.interaction isKindOfClass:[TSIncomingMessage class]]);
 
     if (self.hasViewState) {
         return;
     }
     self.hasViewState = YES;
 
-    TSMessage *interaction = (TSMessage *)self.interaction;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: It turns out the `TSMessage.body` can be non-nil (Empty string: `@""`) for incoming attachment messages.  So we need to check for an attachment first.

File: Signal/src/ViewControllers/MessageDetailViewController.swift [hunk 3]
```diff
@@ -30,6 +30,13 @@ class MessageDetailViewController: OWSViewController {
 
     var mediaMessageView: MediaMessageView?
 
+    // See comments on updateTextLayout.
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessageDetailViewController.swift]: I'd start with the comments on `updateTextLayout` because this is non-trivial.

File: Signal/src/ViewControllers/MessageDetailViewController.swift [hunk 4]
```diff
@@ -567,4 +589,93 @@ class MessageDetailViewController: OWSViewController {
...
+    //
+    // Therefore we're doing something unusual.  
+    //
+    // * We use an empty UIView "messageTextProxyView" as a placeholder for the
+    //   the UITextView.  It has the size and position of where the UITextView
+    //   would be normally.
+    // * We use a UITextView inside that proxy that is just large enough to
+    //   render the content onscreen. We then move it around within the proxy
+    //   bounds to render the parts of the proxy which are onscreen.
+    private func updateTextLayout() {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessageDetailViewController.swift]: I've tested and this works well on iPhone 5C and can handle very large oversize text messages.  It's pretty smooth visually, too. I was afraid there might be some rendering lag.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m (hunk 1 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: :+1:

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 5]
```diff
@@ -336,70 +331,81 @@ - (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
...
+- (nullable TSAttachment *)firstAttachmentIfAnyOfMessage:(TSMessage *)message
+{
+    if (message.attachmentIds.count == 0) {
+        return nil;
+    }
+    NSString *_Nullable attachmentId = message.attachmentIds.firstObject;
+    if (attachmentId.length == 0) {
+        return nil;
+    }
+    return [TSAttachment fetchObjectWithUniqueID:attachmentId];
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: This is kind of a sneaky spot to create a transaction. 

And more generally, since the ViewItem is driving UI code, it seems like we should be doing any reads from a transaction parameter passed in from a longLivedRead db connection which hangs on the view controller (e.g. `uiDatabaseConnection`)

This problem isn't new to this PR, but it just occurred to me.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 6]
```diff
@@ -336,70 +331,81 @@ - (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
...
+}
+
 - (void)ensureViewState
 {
-    OWSAssert([self.interaction isKindOfClass:[TSMessage class]]);
+    OWSAssert([self.interaction isKindOfClass:[TSOutgoingMessage class]] ||
+        [self.interaction isKindOfClass:[TSIncomingMessage class]]);
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Is this significant? 😱

File: Signal/src/ViewControllers/MessageDetailViewController.swift (hunk 4 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/MessageDetailViewController.swift]: This is wild. At what file size does this become relevant?

Comment (charlesmchen) [reply Signal/src/ViewControllers/MessageDetailViewController.swift]: The issue shows up quite easily if you paste oversize text.  Neither `UILabel` and `UITextView` are designed to be used inside a `UIScrollView` to render long strings.

Comment (charlesmchen) [reply Signal/src/ViewControllers/MessageDetailViewController.swift]: Incidentally, I didn't come up with this solution, I've seen it used before to work around other forms of the `UIScrollview`-inside-a-`UIScrollview` issue.

**Comment by charlesmchen** [comment]
> My main concern is about driving the UI on a dbConnection which is not long lived, but that's not new to this PR. Feel free to merge and make a new card.

Card made.

> Also, since this proxy-view code is somewhat clever, it's worth considering only activating it when we're looking at payloads large enough where the simple solution doesn't work. But we could always do that later if it proves problematic.

I think we should always do it:

* This isn't a rare edge case, it's easy to hit this rendering issue with long text.
* There's no safe way to predict what length of text of will trigger this issue.
* If there are any issues with this rendering technique, it'd be nice to know about them sooner rather than later. Using it always will help flush them out.
* Although it takes a minute to understand, I hope this solution is safe.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m (hunk 6 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Not really, I just realized that we might someday accidentally use an `OWSMessageCell` for an info or error message.  This assert will quickly expose that class of bug.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m (hunk 5 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Good catch - card made.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m (hunk 6 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: 👍
